src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/CompileExpressionsTests.cs

```diff
@@ -944,5 +946,172 @@ .maxstack  1
                         """);
                 });
         }
+
+        [WorkItem(66109, "https://github.com/dotnet/roslyn/issues/66109")]
+        [Fact]
+        public void SequencePointsMultipleDocuments()
+        {
+            var sourceA =
+@"partial class Program
+{
+    private int x = 1;
+    private void F()
+    {
+    }
+}";
+            var sourceB =
+@"partial class Program
+{
+    private int y = 2;
+    public Program()
+    {
+        F();
+        int z = x + y;
+#line 999
+    }
+}";
+            var comp = CreateCompilation(
+                new[]
+                {
+                    SyntaxFactory.ParseSyntaxTree(sourceA, path: "A.cs", encoding: Encoding.Default),
+                    SyntaxFactory.ParseSyntaxTree(sourceB, path: "B.cs", encoding: Encoding.Default)
+                },
+                options: TestOptions.DebugDll);
+
+            comp.VerifyPdb("""
+                <symbols>
+                  <files>
+                    <file id="1" name="A.cs" language="C#" checksumAlgorithm="SHA1" checksum="09-65-32-19-5F-F8-8A-58-BF-BC-0C-D3-68-2C-2C-7B-15-33-18-E4" />
+                    <file id="2" name="B.cs" language="C#" checksumAlgorithm="SHA1" checksum="A6-5E-D4-C0-57-0C-1A-FD-F5-27-16-E2-AF-A0-41-0C-94-2D-1B-4C" />
+                  </files>
+                  <methods>
+                    <method containingType="Program" name="F">
+                      <customDebugInfo>
+                        <using>
+                          <namespace usingCount="0" />
+                        </using>
+                      </customDebugInfo>
+                      <sequencePoints>
+                        <entry offset="0x0" startLine="5" startColumn="5" endLine="5" endColumn="6" document="1" />
+                        <entry offset="0x1" startLine="6" startColumn="5" endLine="6" endColumn="6" document="1" />
+                      </sequencePoints>
+                    </method>
+                    <method containingType="Program" name=".ctor">
+                      <customDebugInfo>
+                        <forward declaringType="Program" methodName="F" />
+                        <encLocalSlotMap>
+                          <slot kind="0" offset="29" />
+                        </encLocalSlotMap>
+                      </customDebugInfo>
+                      <sequencePoints>
+                        <entry offset="0x0" startLine="3" startColumn="5" endLine="3" endColumn="23" document="1" />
+                        <entry offset="0x7" startLine="3" startColumn="5" endLine="3" endColumn="23" document="2" />
```

Comment:
Just to make sure I understood properly. These sequence points refer to documents 1 and 2, and so when we ask for `GetDocumentsForMethod` we get more than one and that was causing the problem.
Is that right?

Also, I assume you've verified the fix manually in the end-to-end scenario (winforms) that was reported?

Reply:
> These sequence points refer to documents 1 and 2, and so when we ask for `GetDocumentsForMethod` we get more than one and that was causing the problem.

The sequence points refer to multiple documents so the single document handle emitted to the PDB is nil - see [https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs#L638-L639](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs#L638-L639). When reading the `MethodDebugInfo` from the PDB, we were assuming the document handle was always set, so the debug info which includes the local scopes was dropped.

> I assume you've verified the fix manually in the end-to-end scenario (winforms) that was reported?

I've verified debugging in a winforms project similar to the example in [https://github.com/dotnet/winforms/issues/8354](https://github.com/dotnet/winforms/issues/8354) before and after this change.